File: core/src/com/alexpi/awesometanks/entities/blocks/Spawner.java
package com.alexpi.awesometanks.entities.blocks;

import com.alexpi.awesometanks.entities.DamageListener;
import com.alexpi.awesometanks.entities.tanks.EnemyTank;
import com.alexpi.awesometanks.utils.Constants;
import com.alexpi.awesometanks.utils.Utils;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.PolygonShape;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.scenes.scene2d.Group;

/**
 * Created by Alex on 15/02/2016.
 */
public class Spawner extends Block {
    private long lastSpawn, interval;
    private final AssetManager manager;
    private final Vector2 targetPosition;
    private final int maxType;
    private final Group entityGroup;

    public Spawner(DamageListener listener, AssetManager manager, Group entityGroup, World world, Vector2 targetPosition, Vector2 pos, int level) {
        super(manager,"sprites/spawner.png", world,new PolygonShape(), getHealth(level), pos,1f, true, listener);
        this.manager = manager;
        this.entityGroup = entityGroup;
        this.targetPosition = targetPosition;
        this.fixture.getFilterData().maskBits = Constants.CAT_PLAYER | Constants.CAT_PLAYER_BULLET | Constants.CAT_ITEM;
        lastSpawn = System.currentTimeMillis();
        interval = 1000;
        maxType = getMaxType(level);
    }

    private static int getHealth(int level){
        return 200 + (800-200)/30*level;
    }

    public static int getMaxType(int level) {
        if(level <=3)
            return Constants.RICOCHET;
        else return Constants.RAILGUN;
    }

    @Override
    public void act(float delta) {
        super.act(delta);
        if(lastSpawn + interval <  System.currentTimeMillis()){
            lastSpawn = System.currentTimeMillis();
            interval = Utils.getRandomInt(10000,15000);
            entityGroup.addActor(new EnemyTank(manager, entityGroup, body.getWorld(), body.getPosition(),targetPosition, .75f, Utils.getRandomInt(maxType+1), getDamageListener() ));
        }
    }
}

File: core/src/com/alexpi/awesometanks/entities/blocks/Spawner.kt
package com.alexpi.awesometanks.entities.blocks

import com.alexpi.awesometanks.entities.components.body.CAT_ITEM
import com.alexpi.awesometanks.entities.components.body.CAT_PLAYER
import com.alexpi.awesometanks.entities.components.body.CAT_PLAYER_BULLET
import com.alexpi.awesometanks.entities.items.GoldNugget
import com.alexpi.awesometanks.entities.tanks.EnemyTank
import com.alexpi.awesometanks.screens.LEVEL_COUNT
import com.alexpi.awesometanks.utils.RandomUtils
import com.alexpi.awesometanks.weapons.Weapon
import com.alexpi.awesometanks.world.ExplosionManager
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.physics.box2d.Shape
import com.badlogic.gdx.utils.TimeUtils
import kotlin.experimental.or

/**
 * Created by Alex on 15/02/2016.
 */
class Spawner(private val explosionManager: ExplosionManager, level: Int, pos: Vector2) : BaseBlock(
    "sprites/spawner.png",
    Shape.Type.Polygon,
    getHealth(level),
    pos,
    1f,
    true,
    true
) {
    private var lastSpawn: Long
    private var interval: Long
    private var maxSpan = 15000
    private var generatedTypes: List<Weapon.Type> = getEnemyTypes(level)
    private val nuggetValue: Int
    override fun onAlive(delta: Float) {
        super.onAlive(delta)

        if (lastSpawn + interval < TimeUtils.millis() && !isFrozen) {
            lastSpawn = TimeUtils.millis()
            //It increments the spawn interval by 5 seconds each time to prevent excessive enemy generation
            interval = RandomUtils.getRandomInt(maxSpan - 5000, maxSpan).toLong()
            maxSpan += 5000
            parent.addActor(
                EnemyTank(
                    explosionManager,
                    body.position,
                    EnemyTank.Tier.NORMAL,
                    generatedTypes.random()
                ).also { it.damageListener = damageListener }
            )
        }
    }

    companion object {
        private fun getHealth(level: Int): Float {
            return 400f + level.toFloat() / (LEVEL_COUNT - 1) * 1000f
        }

        private fun getNuggetValue(level: Int): Int {
            return 60 + (level.toFloat() / (LEVEL_COUNT - 1) * 80).toInt()
        }

        private fun getMaxType(level: Int): Int {
            return if (level <= 7) Weapon.Type.RICOCHET.ordinal
            else if (level <= 15) Weapon.Type.ROCKETS.ordinal
            else if (level <= 22) Weapon.Type.LASERGUN.ordinal
            else Weapon.Type.RAILGUN.ordinal
        }

        private fun getMinType(level: Int): Int {
            return if (level <= 10) Weapon.Type.MINIGUN.ordinal
            else if (level <= 12) Weapon.Type.SHOTGUN.ordinal
            else if (level <= 16) Weapon.Type.RICOCHET.ordinal
            else Weapon.Type.FLAMETHROWER.ordinal
        }

        private fun getEnemyTypes(level: Int): List<Weapon.Type> {
            val minType = getMinType(level)
            val maxType = getMaxType(level)

            return Weapon.Type.values().slice(minType..maxType)
        }
    }

    override fun onDestroy() {
        dropLoot()
        super.onDestroy()
    }

    private fun dropLoot() {
        val num1 = RandomUtils.getRandomInt(10, 15)
        for (i in 0 until num1)
            parent.addActor(
                GoldNugget(
                    body.position,
                    RandomUtils.getRandomInt(nuggetValue - 10, nuggetValue + 10)
                )
            )
    }

    init {
        fixture.filterData.maskBits =
            (CAT_PLAYER or CAT_PLAYER_BULLET or CAT_ITEM)
        lastSpawn = TimeUtils.millis()
        interval = 1000
        nuggetValue = getNuggetValue(level)
    }
}